File: programs/marketplace/src/instructions/create_market.rs
use crate::auth::check_is_admin;
use crate::events::{
    emit_market_created_event, emit_market_liquidity_price_event, emit_market_outcome_price_events,
};
use crate::state::market::Market;
use anchor_lang::prelude::*;

#[derive(Accounts)]
pub struct CreateMarket<'info> {
    #[account(init, payer = creator, space = Market::LEN)]
    pub market: Account<'info, Market>,
    #[account(mut)]
    pub creator: Signer<'info>,
    pub system_program: Program<'info, System>,
}

#[access_control(check_is_admin(&ctx.accounts.creator.key))]
pub fn create_market(
    ctx: Context<CreateMarket>,
    name: String,
    about: String,
    category: String,
    image_url: String,
    fee_percentage: f64,
    resolution_source: String,
    resolver: String,
    expires_at: i64,
    expected_value: String,
    resolution_operator: String,
) -> Result<()> {
    let market: &mut Account<Market> = &mut ctx.accounts.market;
    let creator: &Signer = &ctx.accounts.creator;

    market.init(
        creator,
        name,
        about,
        category,
        image_url,
        fee_percentage,
        resolution_source,
        resolver,
        expires_at,
        expected_value,
        resolution_operator,
    )?;

    emit_market_created_event(market, creator);
    emit_market_liquidity_price_event(market);
    emit_market_outcome_price_events(market);

    Ok(())
}

File: app/src/api/createMarket.ts
import {getWorkspace} from "context/workspace";
import * as anchor from "@project-serum/anchor";
import {Program} from "@project-serum/anchor";
import {LAMPORTS_PER_SOL, Transaction} from "@solana/web3.js";
import {BN} from "bn.js";
import {DateTime} from "luxon";

export async function createMarket(marketInfo: any) {
    const workspace = getWorkspace();
    if (!workspace.isReady) {
        return;
    }
    const program = workspace.program as Program;

    const marketKeypair = anchor.web3.Keypair.generate();
    const shareKeyPair = anchor.web3.Keypair.generate();
    const authority = (program.provider as anchor.AnchorProvider).wallet;

    const expiresAt = DateTime.fromJSDate(marketInfo.expiresAt);

    const createMarketInstruction = await program.methods
        .createMarket(
            marketInfo.name,
            marketInfo.about,
            marketInfo.category,
            marketInfo.imageUrl,
            marketInfo.feePercentage,
            marketInfo.resolutionSource,
            marketInfo.resolver,
            new anchor.BN(expiresAt.toUnixInteger()),
            marketInfo.expectedValue,
            marketInfo.resolutionOperator
        )
        .accounts({
            market: marketKeypair.publicKey,
            creator: authority.publicKey
        })
        .signers([marketKeypair])
        .instruction();

    const initShareInstruction = await program.methods
        .initShare()
        .accounts({
            share: shareKeyPair.publicKey,
            market: marketKeypair.publicKey,
            authority: authority.publicKey,
        })
        .signers([shareKeyPair])
        .instruction();

    const addLiquidityInstruction = await program.methods
        .addLiquidity(new BN(marketInfo.initialLiquidity * LAMPORTS_PER_SOL))
        .accounts({
            authority: authority.publicKey,
            market: marketKeypair.publicKey,
            share: shareKeyPair.publicKey,
        })
        .signers([])
        .instruction();

    const transaction = new Transaction();

    transaction.add(createMarketInstruction);
    transaction.add(initShareInstruction);
    transaction.add(addLiquidityInstruction);

    if (marketInfo.bias != 0) {
        const outcome = (marketInfo.bias > 0) ? {yes: {}} : {no: {}};
        const biasAmount = Math.abs(marketInfo.bias);

        const buyOutcomeShareInstruction = await program.methods
            .buyOutcomeShares(outcome, new BN(biasAmount * LAMPORTS_PER_SOL), new BN(0))
            .accounts({
                authority: authority.publicKey,
                share: shareKeyPair.publicKey,
                market: marketKeypair.publicKey,
            })
            .signers([])
            .instruction();

        transaction.add(buyOutcomeShareInstruction);
    }

    transaction.feePayer = authority.publicKey;
    transaction.recentBlockhash = (await workspace.provider!.connection.getLatestBlockhash()).blockhash;

    transaction.sign(marketKeypair, shareKeyPair);

    let tx = await authority.signTransaction(transaction);

    const txId = await workspace.provider?.connection?.sendRawTransaction(tx.serialize());
    if (txId) {
        await workspace.provider?.connection?.confirmTransaction(txId)
    }

    return marketKeypair.publicKey.toBase58();
}
